### PR TITLE
chore(Box|Button|Image|Avatar): converting components with hooks

### PR DESCRIPTION
**Pull request by mnajdova**
This PR converts `Box`, `Button`, `Image`, `Avatar` components to use our newly shipped React hooks as part of https://github.com/microsoft/fluent-ui-react/issues/2269.

- no performance improvements or degrades
- no visual regressions

Adds `FluentComponentStaticProps` type to use Fluent specific static props (`className`, `handledProps`) on function components.

**BREAKING CHANGES**
We are now restricting the set of props provided to the styles. Here are the props propagated to the updated components;
- Button: `text`, `primary`, `disabled`, `circular`, `size`, `loading`, `inverted`, `iconOnly`, `fluid` and `hasContent`
- Avatar: `size`
- Image: `avatar`, `circular`, `fluid`

### CODE REVIEW

File: packages/react-bindings/src/telemetry/useTelemetry.ts
Comment (dzearing) [inline packages/react-bindings/src/telemetry/useTelemetry.ts]: I think we should consider a pattern for hooks return values:

If you are returning a value and a setter, use an array to return a double. The problem with arrays is that they convey nothing specific about the return values - no names, no types, no quantity. Just things. If people see an array returned from a hook in a debugger (especially a vanilla js dev), they will assume [ value, setter ].

If you are returning arbitrary things, use an object. You can return any number of type safe things. You can see clear names that make sense in destructuring. It isn't order specific, and you can more easily deprecate things if needed later down the road.

```tsx
return { setStart, setEnd }
```

File: packages/react/src/components/Avatar/Avatar.tsx
Comment (dzearing) [inline packages/react/src/components/Avatar/Avatar.tsx:13]: Avatar.tsx should not refer to fela specifics directly. @jdhuntington for heads up here.

We should expect the base component to simply take in a `classes` prop. So, Avatar renders DOM and injects the right classes.

Then createComponent can abstract how classes gets injected.

```tsx
export const Avatar = createComponent(AvatarBase, { stuff to build `classes` attribute });
```

Comment (dzearing) [reply packages/react/src/components/Avatar/Avatar.tsx:13]: What do you think?

File: packages/react-bindings/src/telemetry/useTelemetry.ts
Comment (mnajdova) [reply packages/react-bindings/src/telemetry/useTelemetry.ts]: Agreed, will update it. @layershifter let's to this with the other hooks too.

File: packages/react/src/components/Avatar/Avatar.tsx
Comment (layershifter) [reply packages/react/src/components/Avatar/Avatar.tsx:13]: It's not Fela specifics. We use this context only by historical reasons. We should use own context instead.

File: packages/react-bindings/src/hooks/useStyles.ts
Comment (layershifter) [inline packages/react-bindings/src/hooks/useStyles.ts]: We don't need this anymore, `animation` prop was removed

File: packages/react-bindings/src/hooks/useStyles.ts
Comment (layershifter) [inline packages/react-bindings/src/hooks/useStyles.ts]: Let's remove caching for now

File: packages/react-proptypes/src/index.ts
Comment (layershifter) [inline packages/react-proptypes/src/index.ts:479]: ⚠️ This change is correct as all behaviors are functions

File: packages/react/src/components/Button/Button.tsx
Comment (layershifter) [inline packages/react/src/components/Button/Button.tsx]: Should it be before `hasChildren`?

File: packages/react-bindings/src/telemetry/useTelemetry.ts
Comment (mnajdova) [reply packages/react-bindings/src/telemetry/useTelemetry.ts]: The hook was moved to a separate PR, and all PRs which don't follow the [value, setter] rule are returning objects: https://github.com/microsoft/fluent-ui-react/pull/2257

File: packages/react/src/components/Button/Button.tsx
Comment (layershifter) [inline packages/react/src/components/Button/Button.tsx]: I suggest a small refactor there to make it more readable and avoid useless functions:

```tsx
 const iconElement = Icon.create(icon, {
    defaultProps: () =>
      getA11Props('icon', {
        styles: resolvedStyles.icon,
        xSpacing: !content ? 'none' : iconPosition === 'after' ? 'before' : 'after',
      }),
  })

  return (
    <ElementType
      {...rtlTextContainer.getAttributes({ forElements: [children] })}
      {...getA11Props('root', {
        onClick: handleClick,
        disabled,
        className: classes.root,
        onFocus: handleFocus,
        ...unhandledProps,
      })}
    >
      {hasChildren ? (
        children
      ) : (
        <>
          {loading &&
            Loader.create(loader || {}, {
              defaultProps: () =>
                getA11Props('loader', {
                  role: undefined,
                  styles: resolvedStyles.loader,
                }),
            })}
          {iconPosition !== 'after' && iconElement}
          {Box.create(content, {
            defaultProps: () =>
              getA11Props('content', { as: 'span', styles: resolvedStyles.content }),
          })}
          {iconPosition === 'after' && iconElement}
        </>
      )}
    </ElementType>
  )
```

File: packages/react/src/types.ts
Comment (layershifter) [inline packages/react/src/types.ts:20]: I don't have better naming ideas

File: packages/react-bindings/src/hooks/useStyles.ts
Comment (mnajdova) [reply packages/react-bindings/src/hooks/useStyles.ts]: Oooookay... :D

File: packages/react/src/components/Button/Button.tsx
Comment (mnajdova) [reply packages/react/src/components/Button/Button.tsx]: yes!

File: packages/react/src/themes/teams/components/Avatar/avatarStyles.ts
Comment (jurokapsiar) [inline packages/react/src/themes/teams/components/Avatar/avatarStyles.ts:34]: isn't this a breaking change?

Comment (mnajdova) [reply packages/react/src/themes/teams/components/Avatar/avatarStyles.ts:34]: added breaking change changelog entry

File: packages/react/src/components/Button/Button.tsx
Comment (miroslavstastny) [inline packages/react/src/components/Button/Button.tsx]: This is extremely fragile and (currently) wrong.
What if I forget to add telemetry to my component? There should be a conformance test for that.

But the main problem I see is that nothing prevents you from calling `setStart` too late or `setEnd` too early.

In this case the whole return is down outside of the telemetry scope - all calls to `getAttributes()`, `getA11Props()`, `renderLoader()`, `renderIcon()`... - which makes the numbers useless :-(

So we should either give up and remove fluent-ui telemetry (no!) or rethink and rework how we measure it (something like a wrapping function for the whole render? And named something like `renderComponent`?)

Comment (mnajdova) [reply packages/react/src/components/Button/Button.tsx]: No `renderComponent` please, we are trying to move out from it. This is an internal thing and we can track it (like you can block the PR because it's not in the Avatar, and Image components :D - oops will ad it) Even when this was in renderComponent, it was setting end even before invoking render() which may be even worse.. Anyway, I will add the telemetry to the other components before merging. Do you have some other proposal for the setting of the start and end?

Comment (miroslavstastny) [reply packages/react/src/components/Button/Button.tsx]: > Even when this was in renderComponent, it was setting end even before invoking render() which may be even worse..

That's not true. That's what @layershifter broke in #1980

Comment (mnajdova) [reply packages/react/src/components/Button/Button.tsx]: Fixed renderComponent, fixed components' telemetry to be accurate. It is used internally now, so we can make sure we are adding it properly until we have better idea of how to improve it.

**Review by layershifter — APPROVED** [review]

**Pull request merged**
